### ISSUE TITLE
BuildAsync + AdamOptimizer mode-collapse on byte-LM Transformer (residual after #1331 + #1364)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1380]
## Summary

`AiModelBuilder.BuildAsync()` paired with `AdamOptimizer.Optimize()` produces uniform-distribution predictions (top-1 = 0%, perplexity = vocab_size = 256) on a byte-LM Transformer config that trains correctly via the per-sample `model.Train(x, y)` path.

This is the residual of [#1331](https://github.com/ooples/AiDotNet/issues/1331) and [#1364](https://github.com/ooples/AiDotNet/pull/1364) — those PRs merged the H5 (parameter-gradient ordering parity) and H6 (two-Adam-impl divergence) fixes upstream, but on AiDotNet 0.204.0 the underlying byte-LM mode-collapse persists.

## Versions

- AiDotNet: **0.204.0**
- AiDotNet.Tensors: **0.81.3**
- AiDotNet.Native.OpenBLAS: **0.81.3**
- .NET: net10.0 / net471 (verified both)
- Date verified: 2026-05-18

## Reproducer

HarmonicEngine `Phase_PAPER_A_PathB_SingleSeed_Runner.Run_single_seed_0` produces (single seed, V=256, dModel=64, heads=2, layers=1, ctxLen=64, epochs=3, lr=1e-3, ~9216 training samples):

| Path | top-1 | top-5 | perplexity | BPC |
|---|---|---|---|---|
| `AiModelBuilder.BuildAsync()` | **0.0%** | 100.0% | **256.00** (uniform) | 8.00 |
| Per-sample `model.Train(x, y)` | **55.7%** | 79.5% | **6.77** | 2.76 |

Same `Transformer<float>` model instance, same `AdamOptimizer<float, Tensor<float>, Tensor<float>>(null, adamOpts)`, same `InMemoryDataLoader<float, Tensor<float>, Tensor<float>>(xTrain, yTrain)`. The only difference is the training-driver path.

## Code paths under test

**Facade path (broken — top1=0%):**

```csharp
var loader = new InMemoryDataLoader<float, Tensor<float>, Tensor<float>>(xTrain, yTrain);
var result = await builder
    .ConfigureModel(model)
    .ConfigureOptimizer(optimizer)
    .ConfigureDataLoader(loader)
    .BuildAsync();
// result.Predict(...) returns uniform distribution
```

**Per-sample path (works — top1=55.7%):**

```csharp
model.SetTrainingMode(true);
foreach (int idx in shuffledOrder) {
    var x = TakeRow(xTrain, idx);   // [1, ctxLen]
    var y = TakeRow(yTrain, idx);   // [1, vocabSize]
    model.Train(x, y);              // direct per-sample
}
model.SetTrainingMode(false);
// model.Predict(...) returns trained distribution
```

## Diagnostic hypothesis

`AiModelBuilder.BuildSupervisedInternalAsync` likely dispatches to `nn.Train(xTrain, yTrain)` with the full stacked tensors after `nn.TrainWithTape`. The downstream `Optimizer.Step` either:

1. Computes gradients on a flattened parameter view that doesn't match the per-layer parameter layout this Transformer chain expects, OR
2. The batched-tensor path runs a different code branch than the per-sample path, and that branch has a residual `m`/`v` state initialization or sign issue specific to byte-LM (after #1331's flat-ordering parity fix)

A diagnostic experiment that would localize the cause: instrument both paths to log `optimizer._m.Sum()`, `_v.Sum()`, and `parameters.Sum()` after each Train call. If the per-sample path produces non-zero moving averages while the BuildAsync path keeps them at zero, the dispatcher is the bug; otherwise it's a downstream forward pass issue.

## Reference

- HarmonicEngine [`tests/Learning/Support/AiDotNetFacadeTransformerLMPredictor.cs`](https://github.com/ooples/HarmonicEngine/blob/research/ledger-overhaul-and-l4-reification/tests/Learning/Support/AiDotNetFacadeTransformerLMPredictor.cs) has the workaround documented inline (commit b908c18).
- Related prior issue: #1331 (closed), PR #1364 (merged).

## Impact

Blocks consumers from using the facade end-to-end for any Transformer LM training. HarmonicEngine carries an explicit per-sample workaround pinned to this issue.

**Comment by ooples** [comment]
**Regression — bug persists after closure as 'COMPLETED' on 2026-05-19.**

Reproducer added to HarmonicEngine: `tests/Learning/Phase_PAPER_A_BuildAsyncDiagnostic_1380Verify.BuildAsync_Transformer_ByteLM_Escapes_Uniform_On_50KB`

**Run on AiDotNet 0.206.0 + AiDotNet.Tensors 0.81.8 + AiDotNet.Native.OpenBLAS 0.81.8** (2026-05-20):

```
BuildAsync OK — 329984 trainable params, train_wall = 169.4s
BuildAsync classical Transformer @ WT2 50KB / 2ep / 1 seed: top-1 = 0.00% (n_eval = 1000)
  uniform floor 1/V = 0.39%
  per-sample bypass reference (PathB Pareto 8-seed mean): top-1 = 11.98% ± 4.88%
```

**The bypass that the HE-side predictors use as a workaround for this issue still produces a (barely) training model (top-1 = 11.98% ± 4.88%) on the same architecture / hyperparameters / corpus / seed, while `BuildAsync` produces uniform output (top-1 = 0.00%, equal to 1/V floor 0.39%).**

## Reproducer details (exact PathB smoke architecture)

- Transformer: `TransformerArchitecture<float>` (encoder=2, heads=2, dModel=128, dFf=256, ctx=64, vocab=256, dropout=0.0, sinusoidal PE, sequence-classification task)
- Loss: `CategoricalCrossEntropyLoss<float>`
- Optimizer: `AdamOptimizer<float, Tensor<float>, Tensor<float>>` with `AdamOptimizerOptions { InitialLearningRate = 1e-3, MaxIterations = 2, UseAdaptiveLearningRate = false }`
- Data: `DataLoaders.FromTensors(xTrain, yTrain)` with `xTrain : [nTrain, 64]` byte IDs as floats and `yTrain : [nTrain, 256]` one-hot targets; nTrain = 5760 (50KB byte-LM corpus, train_stride=8, 90/10 train/val)
- Build pipeline: `builder.ConfigureModel(model).ConfigureOptimizer(optimizer).ConfigureDataLoader(loader).BuildAsync().GetAwaiter().GetResult()`

`BuildAsync` returns a non-null `AiModelResult` with `TotalTrainableParameters = 329984`, train_wall = 169.4s (reasonable for 2 epochs over 5760 samples on the test hardware) — so the build path is exercising training and updating parameters. But the resulting model's `Predict(x)` produces argmax = same byte for every input position → top-1 = 0.00% over 1000 random eval positions (= near-uniform softmax output, as in the original #1380 symptom).

## What changed since closure (or didn't)

The `H5` + `H6` merges (#1331, #1364) plus whatever PR closed this issue are not enough to fix the byte-LM Transformer case. Either:

1. The closing PR addressed a different reproducer / model type (e.g., regression-task tabular nets), and the byte-LM Transformer path has a residual bug that wasn't covered, OR
2. There's a regression in a more recent commit (most plausible: the AdamW switchover, plan-cache changes, or `Configure*` wiring changes in 0.205/0.206)

## Workaround in HarmonicEngine (until this is genuinely resolved)

`AiDotNetFacadeTransformerLMPredictor.cs` line 251 explicitly bypasses BuildAsync:
```csharp
_ = builderObj; // ConfigureXxx fluent surface intentionally unused
                // until BuildAsync's residual mode-collapse on
                // this Transformer config is fixed upstream.
```
…and uses a per-sample `model.Train(sample, target)` loop instead. This produces meaningful (if still mode-collapse-prone) training: PathB Pareto 8-seed mean top-1 = 11.98% ± 4.88%.

## Ask

1. Re-open this issue.
2. Run the HE-side reproducer (or copy the code from this comment into an AiDotNet integration test) to confirm reproduction on the maintainer's hardware.
3. If reproduced: bisect 0.198 → 0.206 to identify the regressing commit, OR investigate why the original 2026-05-19 closing PR didn't actually fix this code path.
4. Once truly resolved, HE will re-enable BuildAsync + the full `Configure*` surface (`ConfigureMixedPrecision`, `ConfigureJitCompilation`, `ConfigureGpuAcceleration`, `ConfigureRegularization`) which we cannot exercise while bypassing BuildAsync. The HE-side regression test (`Phase_PAPER_A_BuildAsyncDiagnostic_1380Verify`) will catch any future drift.

**Comment by ooples** [comment]
## Deeper diagnostic (2026-05-20, continuation)

Attempted fix: collapsed the batched-stack-then-train path in AiModelBuilder.BuildSupervisedInternalAsync to always per-sample (lines 2241-2264, now always-per-sample). **Result: still mode-collapses.**

Updated regression gate at `Phase_PAPER_A_BuildAsyncDiagnostic_1380Verify` now double-evals via BOTH `trained.Predict` (AiModelResult wrapper) AND `model.Predict` (bare Transformer instance) to bisect where the bug lives:

```
BuildAsync OK -- 329984 trainable params, train_wall = 160.0s
  pathA trained.Predict (AiModelResult wrapper): top-1 = 0.00%
  pathB model.Predict   (bare Transformer):      top-1 = 0.00%
  uniform floor 1/V = 0.39%
  per-sample bypass reference: top-1 = 11.98% ± 4.88%
```

**A=B=0%** means the bug is in training itself, NOT in the AiModelResult wrapper (preprocessing / postprocessing / safety filter / JIT cache). The underlying Transformer's parameters DID update for 160s of training but produced argmax-constant predictions.

## Smoking-gun hypothesis

Lines 2219-2223 of AiModelBuilder.cs:
```csharp
if (_optimizer is IGradientBasedOptimizer<T, Tensor<T>, Tensor<T>> gboTrain
    && nn is NeuralNetworks.NeuralNetworkBase<T> nnBaseForOpt)
{
    nnBaseForOpt.SetBaseTrainOptimizer(gboTrain);
}
```

This replaces the model's default optimizer with the user-supplied one. **HE's working bypass does NOT call SetBaseTrainOptimizer** — it lets the model use its own internal optimizer (constructed lazily via `GetOrCreateBaseOptimizer`). When the bypass + per-sample loop runs, top-1 = 11.98%. When BuildAsync + per-sample loop runs (same arch, same data, same hyperparameters), top-1 = 0%.

Hypothesis: `SetBaseTrainOptimizer` is wiring the user's optimizer in a way that prevents the gradient from reaching the encoder (or the optimizer's internal state is being reset between calls, or the parameter list bound at SetBaseTrainOptimizer time doesn't match the model's actual params after the first forward+backward).

## Suggested next bisect for the maintainer

1. Try BuildAsync WITHOUT SetBaseTrainOptimizer (lazy-construct optimizer via model's default). If that works, the bug is in how the user-supplied optimizer is wired.
2. Try BuildAsync WITH SetBaseTrainOptimizer but the SAME optimizer instance the model would have lazy-constructed. If that works, the bug is in the user-supplied optimizer's initial state (untrained / wrong parameter list).
3. dotnet-trace the gradient norms across the first ~10 Train calls under both paths (bypass + BuildAsync). The diverging step should be visible.

## HE-side status

Reverting the local 0.207.0-fix1380 pin back to public 0.206.0. HE continues to use the per-sample bypass at `AiDotNetFacadeTransformerLMPredictor.cs` line 251. The STAR ledger row at commit 1684cca holds — the classical baseline is "per-sample bypass training" not "properly-Configure*-trained Transformer", and that limitation is now explicit in mechanism_failures.md. Rerunning the regression gate any time after the upstream fix lands will auto-validate.
